packages/slate-react/src/plugin/react-editor.ts

```diff
@@ -778,6 +793,35 @@ export const ReactEditor: ReactEditorInterface = {
             offset -= el.textContent!.length
           })
         }
+      } else if (nonEditableNode) {
+        // Find the edge of the nearest leaf in `searchDirection` from the
+        // non-editable node
+        const leafNodes = Array.from(
+          editorEl.querySelectorAll(
+            // This editor's leaves (not those of a nested editor)
+            '[data-slate-leaf]:not(:scope [data-slate-editor] [data-slate-leaf])'
+          )
+        )
+        leafNode =
+          (searchDirection === 'forward'
+            ? leafNodes.find(leaf => isAfter(nonEditableNode, leaf))
+            : leafNodes.findLast(leaf => isBefore(nonEditableNode, leaf))) ??
+          null
```

Comment:
I've thought of a possible optimization here. Currently, `leafNodes` is a collection of all leaf nodes in the entire editor. But since all Slate elements contain leaves, the furthest we should have to search from the `nonEditableNode` is one element away. I'm going to edit this so that `leafNodes` contains only the leaf nodes in the current Slate element and the element next to it in `searchDirection`.

Reply:
Done in https://github.com/ianstormtaylor/slate/pull/5716/commits/9534c218343c72f535c45af5718037c8c42d9e89